isearch-mb.el

```diff
@@ -127,56 +156,71 @@
+    (with-minibuffer-selected-window
+      (apply args)
+      (unless isearch-mode
+        (throw 'isearch-mb--continue '(ignore))))))
 
 (defun isearch-mb--after-exit (&rest args)
   "Evaluate ARGS, after quitting Isearch-Mb.
 Intended as an advice for commands that quit Isearch and use the
 minibuffer."
-  (if isearch-mb--session
-      (throw 'isearch-mb--continue args)
-    (apply args)))
+  (throw 'isearch-mb--continue args))
 
 (defun isearch-mb--session ()
   "Read search string from the minibuffer."
   (condition-case nil
       (apply
        (catch 'isearch-mb--continue
-         (cl-letf ((isearch-mb--session t)
-                   ((cdr isearch-mode-map) nil)
+         (cl-letf (((cdr isearch-mode-map) nil)
                    ;; We need to set `inhibit-redisplay' at certain points to
                    ;; avoid flicker.  As a side effect, window-start/end in
                    ;; `isearch-lazy-highlight-update' will have incorrect values,
                    ;; so we need to lazy-highlight the whole buffer.
                    (lazy-highlight-buffer (not (null isearch-lazy-highlight))))
            (minibuffer-with-setup-hook
                (lambda ()
-                 (add-hook 'after-change-functions 'isearch-mb--after-change nil 'local)
-                 (add-hook 'post-command-hook 'isearch-mb--post-command-hook nil 'local)
+                 (add-hook 'after-change-functions #'isearch-mb--after-change nil 'local)
+                 (add-hook 'post-command-hook #'isearch-mb--post-command-hook nil 'local)
                  (setq-local tool-bar-map isearch-tool-bar-map)
                  (setq isearch-mb--prompt-overlay (make-overlay (point-min) (point-min)
                                                                 (current-buffer) t t))
                  (isearch-mb--update-prompt)
                  (isearch-mb--post-command-hook))
-             (read-from-minibuffer
-              "I-search: "
-              nil
-              isearch-mb-minibuffer-map
-              nil
-              (if isearch-regexp 'regexp-search-ring 'search-ring)
-              (thread-last '(region url symbol sexp line) ;; TODO: make customizable
-                (mapcar 'thing-at-point)
-                (delq nil)
-                (delete-dups)
-                (mapcar (if isearch-regexp 'regexp-quote 'identity)))
-              t))
+             (unwind-protect
+                 (progn
+                   ;; TODO move advice installation to extra function?
+                   (dolist (fun isearch-mb--after-exit)
+                     (advice-add fun :around #'isearch-mb--after-exit))
+                   (dolist (fun isearch-mb--with-buffer)
+                     (advice-add fun :around #'isearch-mb--with-buffer))
+                   (advice-add #'isearch--momentary-message :override #'isearch-mb--momentary-message)
+                   (advice-add #'isearch-pre-command-hook :override #'isearch-mb--disable-pre-command-hook)
```

Comment:
Instead of overriding the pre command hook it may makes sense to simply remove it and add it back afterwards?